### PR TITLE
RELATED: RAIL-2800 Add placeholders documentation

### PR DESCRIPTION
**Pull request by kandl**
JIRA: RAIL-2800

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-ui-documentation-validation/728/ : SUCCESS in 1m 57s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-oss-check/462/ : SUCCESS in 8s

File: website/i18n/en.json [hunk 1]
```diff
@@ -245,6 +245,10 @@
         "title": "Filter Visual Components",
         "sidebar_label": "Filter Visual Components"
       },
+      "version-8.4.0-placeholders": {
```
Comment (no23reason) [inline website/i18n/en.json:248]: Please revert this also.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-ui-documentation-validation/729/ : SUCCESS in 2m 22s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-oss-check/463/ : SUCCESS in 10s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-ui-documentation-validation/730/ : SUCCESS in 1m 09s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-oss-check/464/ : SUCCESS in 7s

File: docs/30_tips__placeholders.md [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+                Revenue
+            </button>
+        <div/>
+    );
+};
+```
+
+## Common placeholders
+
+Common placeholders may hold one or multiple values (attributes/measures/filters/sorts/totals). Note that you should not set other placeholders as the placeholder value - to combine multiple placeholders, use composed placeholder instead. You can create common placeholder by calling `newPlaceholder` method from `@gooddata/sdk-ui`.
```
Comment (no23reason) [inline docs/30_tips__placeholders.md:82]: Please link to the Composed section when talking about the composed placeholders.

File: docs/30_tips__placeholders.md [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+### Set default value of the common placeholder
+
+```
+// Create common placeholder with default value.
+// Note that in this case, type of the placeholder is derived from the default value,
+// so you don't have to explicitly provide it.
+const measurePlaceholder = newPlaceholder(Md.Revenue);
+
+// In this case, you have to provide the typing as [] would be infered as never[].
+const measuresPlaceholder = newPlaceholder<IMeasure[]>([]);
```
Comment (no23reason) [inline docs/30_tips__placeholders.md:103]: The other examples in the docs are in plain JavaScript so this TypeScript specific thing might be confusing. Maybe mention it just in a comment or something?

File: docs/30_tips__placeholders.md [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+const arithmeticMeasurePlaceholder = newComposedPlaceholder(
+    [primaryMeasurePlaceholder],
+    // Second argument offer a way to provide custom context for the resolution
+    ([primaryMeasure], { operator }) => {
+        if (!primaryMeasure) {
+            return;
+        }
+
+        // Create arithmetic measure from the input
+        return newArithmeticMeasure([primaryMeasure], operator);
```
Comment (no23reason) [inline docs/30_tips__placeholders.md:214]: Arithmetic measure with just one operand is rather strange, can you please add another measure?

File: docs/30_tips__placeholders.md [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+};
+```
+
+## Hooks
+
+### Default hook attached to each placeholder
+
+```
+// By default, each placeholder has usePlaceholder hook attached to it for the convenience
+const measurePlaceholder = newPlaceholder<IMeasure>();
```
Comment (no23reason) [inline docs/30_tips__placeholders.md:255]: Please mention the TypeScript-only stuff in a comment only.

File: docs/30_tips__placeholders.md [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+    const result = useComposedPlaceholder(composedPlaceholder);
+
+    ...
+};
+```
+
+### usePlaceholder
+
+```
+const measurePlaceholder = newPlaceholder<IMeasure>();
```
Comment (no23reason) [inline docs/30_tips__placeholders.md:284]: Please mention the TypeScript-only stuff in a comment only.

File: docs/30_tips__placeholders.md [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+        return updatedMeasure;
+    });
+};
+```
+
+### usePlaceholders
+
+```
+const measurePlaceholder = newPlaceholder<IMeasure>();
+const attributePlaceholder = newPlaceholder<IAttribute>();
```
Comment (no23reason) [inline docs/30_tips__placeholders.md:312]: Please mention the TypeScript-only stuff in a comment only.

File: docs/30_tips__placeholders.md [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+
+```
+const measuresPlaceholder = newPlaceholder([Md.Revenue, Md.Costs]);
+
+const selectedMeasurePlaceholder = newComposedPlaceholder(
+    [measuresPlaceholder],
+    // measures is resolved value of measures placeholder.
+    // Second argument is custom resolution context
+    // wich you may provide at the time you are calling it.
+    ([measures], { measureIndex }: { measureIndex: number }) => {
```
Comment (no23reason) [inline docs/30_tips__placeholders.md:346]: Please mention the TypeScript-only stuff in a comment only.

File: docs/30_tips__placeholders.md [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+    // Usage with useComposedPlaceholder hook
+    const measure = useComposedPlaceholder(selectedMeasurePlaceholder, { measureIndex: 0 });
+};
+```
+
+### useResolveValueWithPlaceholders
+
+```
+const measurePlaceholder = newPlaceholder(Md.Revenue);
+const measuresPlaceholder = newPlaceholder([Md.Costs]);
```
Comment (no23reason) [inline docs/30_tips__placeholders.md:367]: Please consider more distinct names for these, I had trouble catching the difference. Something like `singleMeasurePlaceholder` and `measureArrayPlaceholder`?

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-ui-documentation-validation/732/ : SUCCESS in 56s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-oss-check/466/ : SUCCESS in 8s

**Review by no23reason — APPROVED** [review]
LGTM

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-ui-documentation-validation/739/ : SUCCESS in 1m 47s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-oss-check/471/ : SUCCESS in 8s

**Comment by evgeniabogranskaya** [comment]
LGTM

**Review by no23reason — APPROVED** [review]
LGTM

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-gate-PR418

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-ui-documentation-validation/740/ : SUCCESS in 57s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-fossa-scanning-zuul-el7/118/ : FAILURE in 2m 31s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sonarqube-scan-zuul-gate/20/ : SUCCESS in 43s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-merge-compliance-check/209/ : SUCCESS in 18s

**Pull request merged**